File: gen_exp_boiler.py
import argparse
import os
import datetime
import logging
import json
import itertools
import sys
from copy import copy


RAND_SEED = 1337


def cart_prod_dict(d):
    return (dict(zip(d, x)) for x in itertools.product(*d.values()))


def vals_to_str(v):
    if isinstance(v, list):
        return " ".join(vals_to_str(k) for k in v)
    else:
        return str(v)


if __name__ == "__main__":

    parser = argparse.ArgumentParser()

    parser.add_argument("config",
                        help="Path to json config file",
                        type=str)
    parser.add_argument("-d", "--defaults",
                        help="Path to json config file containing defaults",
                        default='defaults.json',
                        type=str)
    parser.add_argument("-s", "--script-name",
                        help="Path to script to gen grid for",
                        default='synth_exp.py',
                        type=str)
    parser.add_argument("--seeds",
                        help="list of seeds to replicate runs",
                        nargs='+',
                        default=None,
                        type=int)
    parser.add_argument("-e", "--env",
                        help="Bash environment string",
                        default='',
                        type=str)

    args = parser.parse_args()

    #
    # reading defaults from json
    defaults = None
    with open(args.defaults) as f:
        defaults = json.load(f)

    #
    # load specific configs
    config = None
    with open(args.config) as f:
        config = json.load(f)

    #
    # cartesian product of dictionaries
    configs = cart_prod_dict(config)

    for exp_id, c in enumerate(configs):

        seeds = None
        if not args.seeds:
            seeds = [RAND_SEED]
        else:
            seeds = args.seeds

        for seed in seeds:

            d = copy(defaults)
            d.update(c)

            #
            # create command string
            cmd = f"{args.env} {d['interpreter']} {args.script_name} {d['expname']}"
            d.pop('interpreter')
            d.pop('expname')

            cmd += f" --exp-id {exp_id} "

            cmd += f" --seed {seed} "

            for k, v in d.items():
                cmd += f" --{k} {vals_to_str(v)} "

            print(cmd)

File: gen_exp_boiler.jl
using ArgParse
using Dates
using JSON
using Logging
using IterTools

RAND_SEED = 1337
DEFALT_CMD = "julia"

function cart_prod_dict(d)
    return Tuple(Dict(zip(keys(d), x)) for x in IterTools.product(values(d)...))
end

function vals_to_str(v)
    if isa(v, Array)
        return join(v, " ")
    else
        return repr(v)
    end
end

function parse_cmd_line()
    s = ArgParseSettings()

    @add_arg_table s begin
        "config"
            help = "Path to config (JSON) file"
            arg_type = String
            required = true
        "--output", "-o"
            help = "Output directory to print commands"
            arg_type = String
            default = "scripts/"
        "--defaults", "-d"
            help = "Path to default json"
            arg_type = String
            default = nothing
        "--bin", "-b"
            help = "Binary name"
            arg_type = String
            default = "exp.jl"
        "--env", "-e"
            help = "Bash environment strings"
            arg_type = String
            default = ""
        "--exp-id"
            help = "Experiment id"
            arg_type = String
            default = "gen-script"
        "--seeds"
            help = "Seeds for random configurations"
            arg_type = AbstractArray{Int}
            default = [1337]
   
    end

    return parse_args(s)
end

function main()
    
    args = parse_cmd_line()
    
    #
    # creating output dirs, if they do not exist
    # keep track of exps via ids or dates
    date_string = Dates.format(now(), "yyyymmdd-HHMMSSs")
    out_path = joinpath(args["output"], "$(args["exp-id"])_$(date_string)")
    mkpath(out_path)

    #
    # setting local logger
    log_path_dir = joinpath(out_path, "logs")
    mkpath(log_path_dir)
    log_path = joinpath(log_path_dir, "log.txt")
    log_io = open(log_path, "w+")
    logger = SimpleLogger(log_io)
    

    #
    # loading default params
    if args["defaults"] != nothing
        defaults = nothing
        open(args["defaults"]) do f
            defaults = JSON.parse(f)
        end
        @info("Loaded defaults: $defaults")
    end

    #
    # check for expected params
    if !haskey(defaults, "cmd")
        defaults["cmd"] = DEFALT_CMD
    end

    #
    # loading exp configs
    config = nothing
    open(args["config"]) do f
       config = JSON.parse(f)
    end
    @info("Loaded configs: $config")

    #
    # cartesian product of dictionaries
    configs = cart_prod_dict(config)

    for (exp_id, c) in enumerate(configs)
        for seed in args["seeds"]
            d = deepcopy(defaults)
            d = merge(d, c)
            @assert haskey(d, "expname")

            #
            # create command string
            cmd = "$(args["env"]) $(d["cmd"]) $(args["bin"]) $(d["expname"])"
            pop!(d, "cmd")
            pop!(d, "expname")

            cmd *= " --exp-id $exp_id "

            cmd *= " --seed $seed "

            for (k, v) in d
                cmd *= " --$(k) $(vals_to_str(v)) "
            end

            println(cmd)

        end
    end
    
end

main()
